### PR TITLE
STORM-2988 Error on initialization of server mk-worker

### PR DESCRIPTION
**Pull request by dbist**
ran tests locally on storm-core and everything is clean.

### CODE REVIEW

File: storm-core/src/jvm/org/apache/storm/metrics2/reporters/JmxStormReporter.java [hunk 1]
```diff
@@ -67,7 +68,7 @@ public void prepare(MetricRegistry metricsRegistry, Map<String, Object> stormCon
     }
 
     public static String getMetricsJMXDomain(Map reporterConf) {
-        return Utils.getString(reporterConf, JMX_DOMAIN);
+        return Utils.getString(reporterConf.get(Config.STORM_DAEMON_METRICS_REPORTER_PLUGIN_DOMAIN), null);
```
Comment (priyank5485) [inline storm-core/src/jvm/org/apache/storm/metrics2/reporters/JmxStormReporter.java]: JMX_DOMAIN used initially is different than Config.STORM_DAEMON_METRICS_REPORTER_PLUGIN_DOMAIN

I think we should stick to JMX_DOMAIN config key unless @ptgoetz  thinks this is good.

Comment (arunmahadevan) [reply storm-core/src/jvm/org/apache/storm/metrics2/reporters/JmxStormReporter.java]: I think user should be able to specify a domain so it cannot be a fixed value like JMX_DOMAIN. So the proposed changes looks fine.

Comment (arunmahadevan) [reply storm-core/src/jvm/org/apache/storm/metrics2/reporters/JmxStormReporter.java]: I think JMX_DOMAIN was intended to be the default value of the domain.

So may be `Utils.getString(reporterConf.get(Config.STORM_DAEMON_METRICS_REPORTER_PLUGIN_DOMAIN), JMX_DOMAIN)`

cc @ptgoetz

Comment (priyank5485) [reply storm-core/src/jvm/org/apache/storm/metrics2/reporters/JmxStormReporter.java]: @arunmahadevan I am not saying we should return that as default value. I am suggesting to use that as the key to look up the conf map.

Comment (dbist) [reply storm-core/src/jvm/org/apache/storm/metrics2/reporters/JmxStormReporter.java]: happy to oblige, would like to draw a consensus as in branch 2.x I can't just import Config class, be it due to moving of the class from store-core to storm-client or something else? It's explicitly looking for the key in the Config class?

Comment (ptgoetz) [reply storm-core/src/jvm/org/apache/storm/metrics2/reporters/JmxStormReporter.java]: @priyank5485 is correct. JMX_DOMAIN is the key for the lookup from the r Porter config.

Comment (ptgoetz) [reply storm-core/src/jvm/org/apache/storm/metrics2/reporters/JmxStormReporter.java]: ‘return Utils.getString(reporterConf, JMX_DOMAIN);’

Should be something like 

‘return Utils.getString(reporterConf.get(JMX_DOMAIN), “localhost”);’

(Apologies, not on a computer atm)

Comment (HeartSaVioR) [reply storm-core/src/jvm/org/apache/storm/metrics2/reporters/JmxStormReporter.java]: `Utils.getString(reporterConf.get(JMX_DOMAIN), null)` would keep the intention and avoid NPE.

**Comment by dbist** [comment]
thank you @priyank5485 @arunmahadevan @ptgoetz and @HeartSaVioR for the lively review, integrated changes, please review

**Comment by arunmahadevan** [comment]
@ptgoetz , any reason to use a different constant while this [config](https://github.com/apache/storm/blob/1.x-branch/storm-core/src/jvm/org/apache/storm/Config.java#L179) exists? And JmxStormReporter seem to use the constants from Config [here](https://github.com/apache/storm/blob/1.x-branch/storm-core/src/jvm/org/apache/storm/metrics2/reporters/JmxStormReporter.java#L41) and [here](https://github.com/apache/storm/blob/1.x-branch/storm-core/src/jvm/org/apache/storm/metrics2/reporters/JmxStormReporter.java#L46) ? 

Is there a documentation to let the users know the right key to use?

**Comment by ptgoetz** [comment]
> @ptgoetz , any reason to use a different constant while this config exists? And JmxStormReporter seem to use the constants from Config here and here ?

@arunmahadevan, to keep them logically separated so either could be potentially deprecated down the road.

**Comment by HeartSaVioR** [comment]
@arunmahadevan @ptgoetz 
Let's "reuse all" or "duplicate all", not reuse some of them partially. And let's leave explanation on JmxStormReporter, since there's no explanation on JmxStormReporter from metrics v2 doc as well as storm.yaml.example.

https://github.com/apache/storm/blob/master/docs/metrics_v2.md
https://github.com/apache/storm/blob/master/conf/storm.yaml.example

**Comment by arunmahadevan** [comment]
We should add docs and an example usage in the storm.yaml.example (which has graphite and console). @dbist will be great if you address this for JMXStormReporter as a part of this PR.

**Comment by ptgoetz** [comment]
@HeartSaVioR Agreed, but those issues can be addressed outside the scope of this bug (i.e. separate JIRA). For now we should focus on getting this fix in since it is currently holding up the release.

**Comment by dbist** [comment]
@ptgoetz you've read my mind, this fix is holding up a release and I can work on the separate jira to address the doc, not a problem. We can target that for next release.

**Comment by dbist** [comment]
doc update jira is https://issues.apache.org/jira/browse/STORM-3062
please review this PR and 2666 for branch 2 fix.

**Comment by ptgoetz** [comment]
+1

**Comment by HeartSaVioR** [comment]
Go ahead. The issue is effectively blocker so I don't want to block this for non-critical thing. Moreover the configurations are already exposed (though it is not documented yet) so someone may have concerned with changing them. 

Ideally we may get a chance to fix it for 2.0, but don't have strong voice to do it.

**Comment by HeartSaVioR** [comment]
The build failure is unrelated to this patch. Build from local clearly succeeds.

**Comment by HeartSaVioR** [comment]
Thanks @dbist I squashed and merged the patch into 1.x-branch via c26e4ad.

Unfortunately, github auto close only works with master branch. Could you please close this issue? Thanks in advance!

**Pull request closed without merging**